src/main/java/appeng/me/storage/MEMonitorIFluidHandler.java

```diff
@@ -0,0 +1,327 @@
+import java.util.concurrent.ConcurrentSkipListMap;
+
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+import net.minecraftforge.fluids.capability.IFluidTankProperties;
+
+import appeng.api.AEApi;
+import appeng.api.config.AccessRestriction;
+import appeng.api.config.Actionable;
+import appeng.api.config.StorageFilter;
+import appeng.api.networking.security.IActionSource;
+import appeng.api.networking.ticking.TickRateModulation;
+import appeng.api.storage.IMEMonitor;
+import appeng.api.storage.IMEMonitorHandlerReceiver;
+import appeng.api.storage.IStorageChannel;
+import appeng.api.storage.channels.IFluidStorageChannel;
+import appeng.api.storage.data.IAEFluidStack;
+import appeng.api.storage.data.IItemList;
+
+
+public class MEMonitorIFluidHandler implements IMEMonitor<IAEFluidStack>, ITickingMonitor
+{
+	private final IFluidHandler handler;
+	private final IItemList<IAEFluidStack> list = AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ).createList();
+	private final HashMap<IMEMonitorHandlerReceiver<IAEFluidStack>, Object> listeners = new HashMap<>();
+	private final NavigableMap<Integer, CachedFluidStack> memory;
+	private IActionSource mySource;
+	private StorageFilter mode = StorageFilter.EXTRACTABLE_ONLY;
+
+	public MEMonitorIFluidHandler( final IFluidHandler handler )
+	{
+		this.handler = handler;
+		this.memory = new ConcurrentSkipListMap<>();
+	}
+
+	@Override
+	public void addListener( final IMEMonitorHandlerReceiver<IAEFluidStack> l, final Object verificationToken )
+	{
+		this.listeners.put( l, verificationToken );
+	}
+
+	@Override
+	public void removeListener( final IMEMonitorHandlerReceiver<IAEFluidStack> l )
+	{
+		this.listeners.remove( l );
+	}
+
+	@Override
+	public IAEFluidStack injectItems( final IAEFluidStack input, final Actionable type, final IActionSource src )
+	{
+		final int filled = this.handler.fill( input.getFluidStack(), type == type.MODULATE );
+
+		if( filled == 0 )
+		{
+			return input.copy();
+		}
+
+		if( type == Actionable.MODULATE )
+		{
+			this.onTick();
+		}
+
+		if( filled == input.getStackSize() )
+		{
+			return null;
+		}
+
+		final IAEFluidStack o = input.copy();
+		o.setStackSize( input.getStackSize() - filled );
+		return o;
+	}
+
+	@Override
+	public IAEFluidStack extractItems( final IAEFluidStack request, final Actionable type, final IActionSource src )
+	{
+		final FluidStack removed = this.handler.drain( request.getFluidStack(), type == Actionable.MODULATE );
+
+		if( removed == null || removed.amount == 0 )
+		{
+			return null;
+		}
+
+		if( type == Actionable.MODULATE )
+		{
+			this.onTick();
+		}
+
+		final IAEFluidStack o = request.copy();
+		o.setStackSize( removed.amount );
+		return o;
+	}
+
+	@Override
+	public IStorageChannel getChannel()
+	{
+		return AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class );
+	}
+
+	@Override
+	public TickRateModulation onTick()
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') The Cyclomatic Complexity of this method "onTick" is 22 which is greater than 10 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)